File: src/main/java/com/integreety/yatspec/e2e/config/FeignInterceptorConfig.java
package com.integreety.yatspec.e2e.config;

import com.integreety.yatspec.e2e.captor.http.RequestCaptor;
import com.integreety.yatspec.e2e.captor.http.ResponseCaptor;
import com.integreety.yatspec.e2e.interceptor.LsdFeignLoggerInterceptor;
import feign.Logger;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.boot.autoconfigure.condition.ConditionalOnClass;
import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
import org.springframework.cloud.openfeign.FeignClientBuilder;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

@Slf4j
@Configuration
@ConditionalOnProperty(name = "yatspec.lsd.db.connectionstring")
@ConditionalOnClass({FeignClientBuilder.class, Logger.Level.class})
@RequiredArgsConstructor
public class FeignInterceptorConfig {

    @Bean
    @ConditionalOnMissingBean(Logger.Level.class)
    public Logger.Level feignLoggerLevel() {
        return Logger.Level.BASIC;
    }

    @Bean
    public LsdFeignLoggerInterceptor lsdFeignLoggerInterceptor(final RequestCaptor requestCaptor, final ResponseCaptor responseCaptor) {
        return new LsdFeignLoggerInterceptor(requestCaptor, responseCaptor);
    }
}

File: feign/src/main/kotlin/io/lsdconsulting/lsd/distributed/interceptor/config/FeignInterceptorConfig.kt
package io.lsdconsulting.lsd.distributed.interceptor.config

import feign.Logger.Level
import io.lsdconsulting.lsd.distributed.interceptor.captor.FeignHttpHeaderRetriever
import io.lsdconsulting.lsd.distributed.interceptor.captor.FeignRequestCaptor
import io.lsdconsulting.lsd.distributed.interceptor.captor.FeignResponseCaptor
import io.lsdconsulting.lsd.distributed.interceptor.captor.common.Obfuscator
import io.lsdconsulting.lsd.distributed.interceptor.captor.http.SourceTargetDeriver
import io.lsdconsulting.lsd.distributed.interceptor.captor.trace.TraceIdRetriever
import io.lsdconsulting.lsd.distributed.interceptor.interceptor.LsdFeignLoggerInterceptor
import io.lsdconsulting.lsd.distributed.interceptor.persistence.RepositoryService
import org.springframework.beans.factory.annotation.Value
import org.springframework.boot.autoconfigure.condition.ConditionalOnClass
import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration

@Configuration
@ConditionalOnProperty(name = ["lsd.dist.connectionString"])
@ConditionalOnClass(Level::class)
class FeignInterceptorConfig {

    @Bean
    @ConditionalOnMissingBean(Level::class)
    fun feignLoggerLevel() = Level.BASIC

    @Bean
    fun lsdFeignLoggerInterceptor(feignRequestCaptor: FeignRequestCaptor, feignResponseCaptor: FeignResponseCaptor) =
        LsdFeignLoggerInterceptor(feignRequestCaptor, feignResponseCaptor)

    @Bean
    fun feignRequestCaptor(
        repositoryService: RepositoryService,
        sourceTargetDeriver: SourceTargetDeriver,
        traceIdRetriever: TraceIdRetriever,
        feignHttpHeaderRetriever: FeignHttpHeaderRetriever,
        @Value("\${spring.profiles.active:#{''}}") profile: String
    ) = FeignRequestCaptor(
        repositoryService, sourceTargetDeriver, traceIdRetriever, feignHttpHeaderRetriever, profile
    )

    @Bean
    fun feignResponseCaptor(
        repositoryService: RepositoryService,
        sourceTargetDeriver: SourceTargetDeriver,
        traceIdRetriever: TraceIdRetriever,
        feignHttpHeaderRetriever: FeignHttpHeaderRetriever,
        @Value("\${spring.profiles.active:#{''}}") profile: String
    ) = FeignResponseCaptor(
        repositoryService, sourceTargetDeriver, traceIdRetriever, feignHttpHeaderRetriever, profile
    )

    @Bean
    fun feignHttpHeaderRetriever(obfuscator: Obfuscator) = FeignHttpHeaderRetriever(obfuscator)
}
